IdentityCore/src/cache/mac/MSIDMacKeychainTokenCache.m

```diff
@@ -712,318 +840,83 @@ - (BOOL)saveAppMetadata:(__unused MSIDAppMetadataCacheItem *)metadata
-                 error:(NSError **)error
+// TODO: To improve the saving logic here (to not pollute keychain)
+- (BOOL)saveAccountMetadata:(MSIDAccountMetadataCacheItem *)item
+                        key:(MSIDAccountMetadataCacheKey *)key
+                 serializer:(id<MSIDExtendedCacheItemSerializing>)serializer
+                    context:(id<MSIDRequestContext>)context
+                      error:(NSError *__autoreleasing *)error
 {
-    assert(jsonObject);
-    assert(serializer);
-    
-    NSData *itemData = [serializer serializeCacheItem:jsonObject];
+    MSIDMacCredentialStorageItem *storageItem = key.isShared ? self.sharedStorageItem : self.appStorageItem;
+    MSIDMacCredentialStorageItem *savedStorageItem = [self storageItemWithKey:key serializer:serializer context:context error:error];
     
-    if (!itemData)
+    if (savedStorageItem)
     {
-        if (error)
-        {
-            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Failed to serialize account item.", nil, nil, nil, context.correlationId, nil);
-        }
-        MSID_LOG_WITH_CTX(MSIDLogLevelError, context, @"Failed to serialize token item.");
-        return NO;
+        [storageItem mergeStorageItem:savedStorageItem inBucket:MSID_ACCOUNT_METADATA_CACHE_TYPE];
     }
     
-    MSID_LOG_WITH_CTX_PII(MSIDLogLevelInfo, context, @"Saving keychain item, item info %@.", MSID_PII_LOG_MASKABLE(jsonObject));
-    
-    return [self saveData:itemData
-                      key:key
-                  context:context
-                    error:error];
+    [storageItem storeItem:item inBucket:MSID_ACCOUNT_METADATA_CACHE_TYPE forKey:key];
+    return [self saveStorageItem:storageItem key:key serializer:serializer context:context error:error];
 }
 
-#pragma mark - Account metadata
-
 - (MSIDAccountMetadataCacheItem *)accountMetadataWithKey:(MSIDAccountMetadataCacheKey *)key
                                               serializer:(id<MSIDExtendedCacheItemSerializing>)serializer
                                                  context:(id<MSIDRequestContext>)context
                                                    error:(NSError *__autoreleasing *)error
 {
-    NSArray<MSIDAccountMetadataCacheItem *> *items =  [self cacheItemsWithKey:key
-                                                                   serializer:serializer
-                                                               cacheItemClass:[MSIDAccountMetadataCacheItem class]
-                                                                      context:context error:error];
+    /*
+     Sync in memory cache with persistent cache at the time of look up.
+     */
+    MSIDMacCredentialStorageItem *storageItem = key.isShared ? self.sharedStorageItem : self.appStorageItem;
+    MSIDMacCredentialStorageItem *savedStorageItem = [self storageItemWithKey:key serializer:serializer context:context error:error];
     
-    if (items.count > 1)
+    if (savedStorageItem)
```

Comment:
this code seems duplicate:
1. get correct storage item
2. read storage item
3. merge storage item from persistence

Can we have one method that does that?

Reply:
Addressed.